processing/src/main/java/io/druid/segment/IndexMergerV9.java

```diff
@@ -0,0 +1,1277 @@
+
+    final List<String> mergedMetrics = IndexMerger.mergeIndexed(
+        Lists.transform(
+            indexes,
+            new Function<IndexableAdapter, Iterable<String>>()
+            {
+              @Override
+              public Iterable<String> apply(@Nullable IndexableAdapter input)
+              {
+                return Iterables.transform(
+                    input.getMetricNames(),
+                    new Function<String, String>()
+                    {
+                      @Override
+                      public String apply(@Nullable String input)
+                      {
+                        return input;
+                      }
+                    }
+                );
+              }
+            }
+        )
+    );
+
+    Function<ArrayList<Iterable<Rowboat>>, Iterable<Rowboat>> rowMergerFn = new Function<ArrayList<Iterable<Rowboat>>, Iterable<Rowboat>>()
+    {
+      @Override
+      public Iterable<Rowboat> apply(
+          @Nullable final ArrayList<Iterable<Rowboat>> boats
+      )
+      {
+        return new MergeIterable<Rowboat>(
+            Ordering.<Rowboat>natural().nullsFirst(),
+            boats
+        );
+      }
+    };
+
+    return makeIndexFiles(indexes, outDir, progress, mergedDimensions, mergedMetrics, null, rowMergerFn, indexSpec);
+  }
+
+  private File makeIndexFiles(
+      final List<IndexableAdapter> adapters,
+      final File outDir,
+      final ProgressIndicator progress,
+      final List<String> mergedDimensions,
+      final List<String> mergedMetrics,
+      final Map<String, Object> segmentMetadata,
+      final Function<ArrayList<Iterable<Rowboat>>, Iterable<Rowboat>> rowMergerFn,
+      final IndexSpec indexSpec
+  ) throws IOException
+  {
+    progress.start();
+    progress.progress();
+
+    final IOPeon ioPeon = new TmpFileIOPeon(false);
+    final FileSmoosher v9Smoosher = new FileSmoosher(outDir);
+    final File v9TmpDir = new File(outDir, "v9-tmp");
+    v9TmpDir.mkdirs();
+    log.info("Start making v9 index files, outDir:%s", outDir);
+
+    ByteStreams.write(
+        Ints.toByteArray(IndexIO.V9_VERSION),
+        Files.newOutputStreamSupplier(new File(outDir, "version.bin"))
+    );
+
+    progress.progress();
+    final Map<String, ValueType> metricsValueTypes = Maps.newTreeMap(Ordering.<String>natural().nullsFirst());
+    final Map<String, String> metricTypeNames = Maps.newTreeMap(Ordering.<String>natural().nullsFirst());
+    final List<ColumnCapabilitiesImpl> dimCapabilities = Lists.newArrayListWithCapacity(mergedDimensions.size());
+    mergeCapabilities(adapters, mergedDimensions, metricsValueTypes, metricTypeNames, dimCapabilities);
+
+    /************* Setup Dim Conversions **************/
+    progress.progress();
+    final Map<String, Integer> dimCardinalities = Maps.newHashMap();
+    final ArrayList<GenericIndexedWriter<String>> dimValueWriters = setupDimValueWriters(ioPeon, mergedDimensions);
```

Comment:
IMO "dimValue" is more consistent with other classes who also treat these as dim values other than dictionary